### PR TITLE
add illumos support

### PR DESCRIPTION
**Pull request by jclulow**
This pull request adds support for building on [illumos](https://illumos.org) systems (e.g., [OmniOS](https://omnios.org/), [OpenIndiana](https://www.openindiana.org/), etc).  I have been using this library with the [`rtdlib` Rust crate](https://github.com/fewensa/fantasy) for a while now on my illumos system.

### CODE REVIEW

**Comment by levlam** [comment]
The pull request looks very promising. It seems that most changes must be applied to build TDLib also on Solaris and aren't specific to illumos. I.e. `CMAKE_SYSTEM_NAME STREQUAL "SunOS"` can be used instead of `ILLUMOS` and `TD_SOLARIS` can be used instead of `TD_ILLUMOS`.

Also, the function `td_set_up_compiler` is supposed to be self-sufficient, so it shouldn't depend on `include(illumos)` in the main CMakeLists.txt, It can detect `illumos` itself, but there are seems to be no reasons to ditinguish it from other SunOS-derived operating systems.

**Comment by levlam** [comment]
Solaris doesn't have epoll and eventfd, so TD_ILLUMOS is still needed, but it would be better to identify it using a define as identified all other operating systems.

**Comment by jclulow** [comment]
> The pull request looks very promising. It seems that most changes must be applied to build TDLib also on Solaris and aren't specific to illumos. I.e. `CMAKE_SYSTEM_NAME STREQUAL "SunOS"` can be used instead of `ILLUMOS` and `TD_SOLARIS` can be used instead of `TD_ILLUMOS`.
> Solaris doesn't have epoll and eventfd, so TD_ILLUMOS is still needed

Right, Oracle have diverged with their proprietary fork over the last ten years.  Support for epoll is just one thing that's different in illumos.  Over time we expect there will be more differences, so we generally feel pretty strongly about not being lumped in with Solaris anymore -- we're a different platform.

> Also, the function `td_set_up_compiler` is supposed to be self-sufficient, so it shouldn't depend on `include(illumos)` in the main CMakeLists.txt, It can detect `illumos` itself, but there are seems to be no reasons to ditinguish it from other SunOS-derived operating systems.

There is at least one divergence: illumos requires `-lnsl -lsocket` whereas, say, Oracle Solaris 11.4 probably doesn't anymore.

I would have used a macro, but unfortunately we don't have one available at present that distinguishes us from Solaris systems; hence the detection logic in `illumos.cmake`.

I have added `include(illumos)` into `td_set_up_compiler()` and tested that it still appears to build correctly -- is that sufficiently self-contained?

**Comment by levlam** [comment]
> There is at least one divergence: illumos requires -lnsl -lsocket whereas, say, Oracle Solaris 11.4 probably doesn't anymore.

If these libraries still exist on Solaris, it should be fine to explicitly link them.

**Comment by levlam** [comment]
> I have added include(illumos) into td_set_up_compiler() and tested that it still appears to build correctly -- is that sufficiently self-contained?

This would be much better.

**Comment by levlam** [comment]
`-ffunction-sections -fdata-sections` also needs to be disabled on illumos, because without linker support of `--gc-sections --exclude-libs,ALL` they only increase binary size.
Is there are any alternative to them for removing unused functions on illumos?

**Review by VChernomYA — APPROVED** [review]

**Comment by VChernomYA** [comment]
authorizationStateWaitTdlibParameters

**Review by VChernomYA — APPROVED** [review]

**Comment by VChernomYA** [comment]
CMake/illumos.cmake

**Review by VChernomYA — COMMENTED** [review]
[]()

**Comment by VChernomYA** [comment]
https://github.com/tdlib/td/blob/eb80924dad30af4e6d8385d058bb7e847174df5e/example/java/org/drinkless/tdlib/example/Example.java#L304

File: tdutils/td/utils/port/uname.cpp [hunk 1]
```diff
@@ -194,7 +194,7 @@ Slice get_operating_system_version() {
     }
 #endif
```
[the remark this answers is not in the document]
Comment (VChernomYA) [reply tdutils/td/utils/port/uname.cpp:195]: Apple iphone 6s

**Review by VChernomYA — APPROVED** [review]

**Comment by jclulow** [comment]
> `-ffunction-sections -fdata-sections` also needs to be disabled on illumos, because without linker support of `--gc-sections --exclude-libs,ALL` they only increase binary size.
> Is there are any alternative to them for removing unused functions on illumos?

We have `-z ignore`, which does something that sounds similar but might end up doing more things.  From [ld(1)](https://illumos.org/man/1/ld):

```
       -z ignore | record

           Ignores, or records, dynamic dependencies that are not referenced
           as part of the link-edit. Ignores, or records, unreferenced ELF
           sections from the relocatable objects that are read as part of the
           link-edit. By default, -z record is in effect.

           If an ELF section is ignored, the section is eliminated from the
           output file being generated. A section is ignored when three
           conditions are true. The eliminated section must contribute to an
           allocatable segment. The eliminated section must provide no global
           symbols. No other section from any object that contributes to the
           link-edit, must reference an eliminated section.
```

I can try and built the library with that option in effect and see how it goes.

As an aside, are the comments from VChernomYA just spam that I can ignore?  It's hard to understand what's going on there...

Cheers.

**Comment by levlam** [comment]
@jclulow

> As an aside, are the comments from VChernomYA just spam that I can ignore?

Yes, comments from VChernomYA are completely unrelated.

> I can try and built the library with that option in effect and see how it goes.

I would like to see the result. If it works as expected, then adding the option will make `illumos` on par with other operating systems. Otherwise the options `-ffunction-sections -fdata-sections` will also need to be disabled.

**Comment by levlam** [comment]
@jclulow In my tests `-z ignore` had no considerable impact on shared library or executable file size.

**Pull request merged**